source/basic_skills.rst

```diff
@@ -886,18 +887,410 @@ The proofs we have carried out in this section provide some hints.
+
+In the first example, applying ``le_trans``
+creates two goals,
+and we use the curly braces to enclose the proof
+of the first one.
+In the fourth example and in the last example, we avoid going into tactic mode entirely:
+``le_trans h₀ h₁`` and ``le_refl x`` are the proof terms we need.
+
+Here are a few more library theorems:
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    variables a b c : ℝ
+
+    -- BEGIN
+    #check (le_refl  : ∀ a, a ≤ a)
+    #check (le_trans : a ≤ b → b ≤ c → a ≤ c)
+    #check (lt_of_le_of_lt : a ≤ b → b < c → a < c)
+    #check (lt_of_lt_of_le : a < b → b ≤ c → a < c)
+    #check (lt_trans : a < b → b < c → a < c)
+    -- END
+
+Use them together with ``apply`` and ``exact`` to prove the following:
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    variables a b c : ℝ
+
+    -- BEGIN
+    example (a b c d e : ℝ) (h₀ : a ≤ b) (h₁ : b < c) (h₂ : c ≤ d)
+        (h₃ : d < e) :
+      a < e :=
+    sorry
+    -- END
+
+In fact, Lean has a tactic that does this sort of thing automatically:
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    variables a b c : ℝ
+
+    -- BEGIN
+    example (a b c d e : ℝ) (h₀ : a ≤ b) (h₁ : b < c) (h₂ : c ≤ d)
+        (h₃ : d < e) :
+      a < e :=
+    by linarith
+    -- END
+
+We will return to ``linarith`` below.
+
+Here are some more theorems in the library that can be used to establish
+inequalities on the real numbers.
+
+.. code-block:: lean
+
+    import analysis.special_functions.exp_log
+    open real
+
+    variables  a b c d : ℝ
+
+    #check (exp_le_exp : exp a ≤ exp b ↔ a ≤ b)
+    #check (exp_lt_exp : exp a < exp b ↔ a < b)
+    #check (log_le_log : 0 < a → 0 < b → (log a ≤ log b ↔ a ≤ b))
+    #check (log_lt_log : 0 < a → a < b → log a < log b)
+    #check (add_le_add : a ≤ b → c ≤ d → a + c ≤ b + d)
+    #check (add_lt_add_of_le_of_lt : a ≤ b → c < d → a + c < b + d)
+    #check (add_lt_add_of_le_of_lt : a ≤ b → c < d → a + c < b + d)
+    #check (add_nonneg : 0 ≤ a → 0 ≤ b → 0 ≤ a + b)
+    #check (add_pos : 0 < a → 0 < b → 0 < a + b)
+    #check (add_pos_of_pos_of_nonneg : 0 < a → 0 ≤ b → 0 < a + b)
+    #check (exp_pos : ∀ a, 0 < exp a)
+
+Some of the theorems, ``exp_le_exp``, ``exp_lt_exp``, and ``log_le_log``
+use a *bi-implication*, which represents the
+phrase "if and only if."
+(You can type it in VS Code with ``\lr`` of ``\iff``).
+We will discuss this connective in greater detail in the next chapter.
+Such a theorem can be used with ``rw`` to rewrite a goal to
+an equivalent one:
+
+.. code-block:: lean
+
+    import analysis.special_functions.exp_log
+
+    open real
+
+    -- BEGIN
+    example (a b : ℝ) (h : a ≤ b) : exp a ≤ exp b :=
+    begin
+      rw [exp_le_exp],
+      exact h
+    end
+    -- END
+
+In this section, however, we will use that fact that if ``h : A ↔ B``
+is such an equivalence,
+then ``h.1`` establishes the forward direction, ``A → B``,
+and ``h.2`` establishes the reverse direction, ``B → A``.
```

Comment:
Do we want to encourage that style? I always prefer `h.mp` and `h.mpr` because they immediately tells readers that `h` is an equivalence, and they are not so much longer. The case of `.left` and `.right` for conjunctions is less clear, both because they are longer and because `.1` and `.2` are closer to the actual semantic.

Reply:
I'll change it to `mp` and `mpr`. I prefer those. I once did a PR to Mathlib and someone (Kenny Lau?) pointed out to me that Mathlib uses `.1` and `.2` much more, so I was trying to fit in. Peer pressure.